schedutils/uclampset.1
.\" uclampset(1) manpage
.\"
.\" Copyright (C) 2020-2021 Qais Yousef <qais.yousef@arm.com>
.\" Copyright (C) 2020-2021 Arm Ltd
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" version 2, as published by the Free Software Foundation.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH UCLAMPSET 1 "August 2020" "util-linux" "User Commands"
.SH NAME
uclampset \- manipulate the utilization clamping attributes of the system or
a process.
.SH SYNOPSIS
.B uclampset
[options]
.RI [ -m\ uclamp_min ]\ [ -M\ uclamp_max ]\ command\  [ argument ...]
.br
.B uclampset
[options]
.RI [ -m\ uclamp_min ]\ [ -M\ uclamp_max ]
.B \-p
.RI pid
.SH DESCRIPTION
.B uclampset
sets or retrieves the utilization clamping attributes of an existing \fIpid\fR,
or runs \fIcommand\fR with the given attributes.

Utilization clamping is a new feature added in v5.3. It gives a hint to the
scheduler about the allowed range of utilization the task should be operating
at.

The utilization of the task affects frequency selection and task placement.
Only schedutil cpufreq governor understands handling util clamp hints at the
time of writing. Consult your kernel docs for further info about other
cpufreq governors support.

If you're running on asymmetric heterogeneous system like Arm's big.LITTLE.
Utilization clamping can help bias task placement. If the task is boosted such
that
.B util_min
value is higher than the little cores' capacity, then the scheduler will do its
best to place it on a big core.

Similarly, if
.B util_max
is smaller than or equal the capacity of the little cores, then the scheduler
can still choose to place it there even if the actual utilization of the task
is at max.

Setting a task's
.B uclamp_min
to a none zero value  will effectively boost the task as when it runs it'll
always start from this utilization value.

By setting a task's
.B uclamp_max
below 1024, this will effectively cap the task as when it runs it'll never be
able to go above this utilization value.

The full utilization range is: [0:1024].
The special value -1 is used to reset to system's default.

.SH OPTIONS
.TP
.B \-m
Set util_min value.
.TP
.B \-M
Set util_max value.
.TP
.BR \-a ,\  \-\-all-tasks
Set or retrieve the utilization clamping attributes of all the tasks (threads)
for a given PID.
.TP
.BR \-p ,\  \-\-pid
Operate on an existing PID and do not launch a new task.
.TP
.BR \-s ,\  \-\-system
Set or retrieve the system-wide utilization clamping attributes.
.TP
.BR \-R ,\  \-\-reset-on-fork
Set
.B SCHED_FLAG_RESET_ON_FORK
flag
.TP
.BR \-v ,\  \-\-verbose
Show status information.
.TP
.BR \-V ,\  \-\-version
Display version information and exit.
.TP
.BR \-h ,\  \-\-help
Display help text and exit.
.SH USAGE
.TP
The default behavior is to run a new command:
.B uclampset
.I [-m\ uclamp_min]
.I [-M\ uclamp_max]
.IR command\  [ arguments ]
.TP
You can also retrieve the utilization clamping attributes of an existing task:
.B uclampset \-p
.I pid
.TP
Or set them:
.B uclampset \-p
.I pid
.I [-m\ uclamp_min]
.I [-M\ uclamp_max]
.TP
Or control the system-wide attributes:
.B uclampset \-s
.I [-m\ uclamp_min]
.I [-M\ uclamp_max]
.SH PERMISSIONS
A user must possess
.B CAP_SYS_NICE
to change the scheduling attributes of a process.  Any user can retrieve the
scheduling information.

.SH NOTES
The system wide utilization clamp attributes are there to control the _allowed_
range the tasks can use. By default both
.B uclamp_min
and
.B uclamp_max
are set to 1024. This means users can set the utilization clamp values for
their task across the full range [0:1024].

.TP
For example:
.B uclampset \-s
.I -m\ 512
.I -M\ 700
.PP
will prevent any task from being boosted higher than 512. And all tasks in the
systems are capped to a utilization of 700. Effectively rendering the maximum
performance of the system to 700.

Consult your kernel docs for the exact expected behavior on that kernel.
.SH AUTHORS
.UR qais.yousef@arm.com
Qais Yousef
.UE
.SH SEE ALSO
.BR nice (1),
.BR renice (1),
.BR taskset (1),
.BR sched (7)
.sp
See
.BR sched_setscheduler (2)
and
.BR sched_setattr (2)
for a description of the Linux scheduling scheme.
.SH AVAILABILITY
The uclampset command is part of the util-linux package and is available from
https://www.kernel.org/pub/linux/utils/util-linux/.


NEWS
util-linux 2.36.1: Nov 16 2020
* see Documentation/releases/v2.36.1-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36.1-ChangeLog

util-linux 2.36: Jul 23 2020
* see Documentation/releases/v2.36-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36-ChangeLog

util-linux 2.36-rc2: Jul 09 2020
* see Documentation/releases/v2.36-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36-rc2-ChangeLog

util-linux 2.36-rc1: Jun 09 2020
* see Documentation/releases/v2.36-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.36/v2.36-rc1-ChangeLog

util-linux 2.35: Jan 21 2020
* see Documentation/releases/v2.35-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.35/v2.35-ChangeLog

util-linux 2.35-rc2: Jan 8 2020
* see Documentation/releases/v2.35-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.35/v2.35-rc2-ChangeLog

util-linux 2.35-rc1: Dec 11 2019
* see Documentation/releases/v2.35-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.35/v2.35-rc1-ChangeLog

util-linux 2.34: Jun 14 2019
* see Documentation/releases/v2.34-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.34/v2.34-ChangeLog

util-linux 2.34-rc2: May 30 2019
* see Documentation/releases/v2.34-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.34/v2.34-rc2-ChangeLog

util-linux 2.34-rc1: Apr 30 2019
* see Documentation/releases/v2.34-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.34/v2.34-rc1-ChangeLog

util-linux 2.33: Nov 6 2018
* see Documentation/releases/v2.33-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-ChangeLog

util-linux 2.33-rc2: Oct 19 2018
* see Documentation/releases/v2.33-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-rc2-ChangeLog

util-linux 2.33-rc1: Sep 25 2018
* see Documentation/releases/v2.33-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-rc1-ChangeLog

util-linux 2.32: Mar 21 2018
* see Documentation/releases/v2.32-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.32/v2.32-ChangeLog

util-linux 2.32-rc2: Mar 01 2018
* see Documentation/releases/v2.32-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.32/v2.32-rc2-ChangeLog

util-linux 2.32-rc1: Feb 13 2018
* see Documentation/releases/v2.32-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.32/v2.32-rc1-ChangeLog

util-linux 2.31: Oct 19 2017
* see Documentation/releases/v2.31-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.31/v2.31-ChangeLog

util-linux 2.31-rc2: Oct 03 2017
* see Documentation/releases/v2.31-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.31/v2.31-rc2-ChangeLog

util-linux 2.31-rc1: Sep 22 2017
* see Documentation/releases/v2.31-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.31/v2.31-rc1-ChangeLog

util-linux 2.30: Jun 02 2017
* see Documentation/releases/v2.30-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.30/v2.30-ChangeLog

util-linux 2.30-rc2: May 12 2017
* see Documentation/releases/v2.30-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.30/v2.30-rc2-ChangeLog

util-linux 2.30-rc1: May 12 2017
* see Documentation/releases/v2.30-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.30/v2.30-rc1-ChangeLog

util-linux 2.29: Nov 8 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-ChangeLog

util-linux 2.29-rc2: Oct 19 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-rc2-ChangeLog

util-linux 2.29-rc1: Sep 30 2016
* see Documentation/releases/v2.29-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.29/v2.29-rc1-ChangeLog

util-linux 2.28: Apr 12 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-ChangeLog

util-linux 2.28-rc2: Mar 29 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-rc2-ChangeLog

util-linux 2.28-rc1: Mar 11 2016
* see Documentation/releases/v2.28-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.28/v2.28-rc1-ChangeLog

util-linux 2.27: Sep 07 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-ChangeLog

util-linux 2.27-rc2: Aug 24 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-rc2-ChangeLog

util-linux 2.27-rc1: Jul 31 2015
* see Documentation/releases/v2.27-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.27/v2.27-rc1-ChangeLog

util-linux 2.26: Feb 19 2015
* see Documentation/releases/v2.26-ReleaseNotes or the complete changelog at
  https://www.kernel.org/pub/linux/utils/util-linux/v2.26/v2.26-ChangeLog

util-linux 2.26-rc2: Feb 4 2015

schedutils/uclampset.c
/*
 * uclampset.c - change utilization clamping attributes of a task or the system
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Copyright (C) 2020-2021 Qais Yousef
 * Copyright (C) 2020-2021 Arm Ltd
 */

#include <errno.h>
#include <getopt.h>
#include <sched.h>
#include <stdio.h>
#include <stdlib.h>

#include "closestream.h"
#include "path.h"
#include "pathnames.h"
#include "procutils.h"
#include "sched_attr.h"
#include "strutils.h"

#define NOT_SET		-2U

struct uclampset {
	unsigned int util_min;
	unsigned int util_max;

	pid_t pid;
	unsigned int	all_tasks:1,		/* all threads of the PID */
			system:1,
			util_min_set:1,		/* indicates -m option was passed */
			util_max_set:1,		/* indicates -M option was passed */
			reset_on_fork:1,
			verbose:1;
	char *cmd;
};

static void __attribute__((__noreturn__)) usage(void)
{
	FILE *out = stdout;

	fprintf(out,
		_(" %1$s [options]\n"
		  " %1$s [options] --pid <pid> | --system | <command> <arg>...\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, out);
	fputs(_("Show or change the utilization clamping attributes of a process or the system.\n"), out);
	fputs(_("Utilization range: [0:1024]\n"), out);
	fputs(_("Use special -1 value to reset to system's default\n"), out);

	fputs(USAGE_OPTIONS, out);
	fputs(_(" -m                   util_min value to set\n"), out);
	fputs(_(" -M                   util_max value to set\n"), out);
	fputs(_(" -a, --all-tasks      operate on all the tasks (threads) for a given pid\n"), out);
	fputs(_(" -p, --pid            operate on existing given pid\n"), out);
	fputs(_(" -s, --system         operate on system\n"), out);
	fputs(_(" -R, --reset-on-fork  set reset-on-fork flag\n"), out);
	fputs(_(" -v, --verbose        display status information\n"), out);

	fputs(USAGE_SEPARATOR, out);
	printf(USAGE_HELP_OPTIONS(22));

	printf(USAGE_MAN_TAIL("uclampset(1)"));
	exit(EXIT_SUCCESS);
}

static void show_uclamp_pid_info(pid_t pid, char *cmd)
{
	struct sched_attr sa;
	char *comm;

	/* don't display "pid 0" as that is confusing */
	if (!pid)
		pid = getpid();

	if (sched_getattr(pid, &sa, sizeof(sa), 0) != 0)
		err(EXIT_FAILURE, _("failed to get pid %d's uclamp values"), pid);

	if (cmd)
		comm = cmd;
	else
		comm = proc_get_command_name(pid);

	printf(_("%s (%d) util_clamp: min: %d max: %d\n"),
	       comm ? : "uknown", pid, sa.sched_util_min, sa.sched_util_max);

	if (!cmd)
		free(comm);
}

static unsigned int read_uclamp_sysfs(char *filename)
{
	unsigned int val;

	if (ul_path_read_u32(NULL, &val, filename) != 0)
		err(EXIT_FAILURE, _("cannot read %s"), filename);

	return val;
}

static void write_uclamp_sysfs(char *filename, unsigned int val)
{
	if (ul_path_write_u64(NULL, val, filename) != 0)
		err(EXIT_FAILURE, _("cannot write %s"), filename);
}

static void show_uclamp_system_info(void)
{
	unsigned int min, max;

	min = read_uclamp_sysfs(_PATH_PROC_UCLAMP_MIN);
	max = read_uclamp_sysfs(_PATH_PROC_UCLAMP_MAX);

	printf(_("System util_clamp: min: %u max: %u\n"), min, max);
}

static void show_uclamp_info(struct uclampset *ctl)
{
	if (ctl->system) {
		show_uclamp_system_info();
	} else if (ctl->all_tasks) {
		pid_t tid;
		struct proc_tasks *ts = proc_open_tasks(ctl->pid);

		if (!ts)
			err(EXIT_FAILURE, _("cannot obtain the list of tasks"));

		while (!proc_next_tid(ts, &tid))
			show_uclamp_pid_info(tid, NULL);

		proc_close_tasks(ts);
	} else {
		show_uclamp_pid_info(ctl->pid, ctl->cmd);
	}
}

static int set_uclamp_one(struct uclampset *ctl, pid_t pid)
{
	struct sched_attr sa;

	if (sched_getattr(pid, &sa, sizeof(sa), 0) != 0)
		err(EXIT_FAILURE, _("failed to get pid %d's uclamp values"), pid);

	if (ctl->util_min_set)
		sa.sched_util_min = ctl->util_min;
	if (ctl->util_max_set)
		sa.sched_util_max = ctl->util_max;

	sa.sched_flags = SCHED_FLAG_KEEP_POLICY |
			 SCHED_FLAG_KEEP_PARAMS |
			 SCHED_FLAG_UTIL_CLAMP_MIN |
			 SCHED_FLAG_UTIL_CLAMP_MAX;

	if (ctl->reset_on_fork)
		sa.sched_flags |= SCHED_FLAG_RESET_ON_FORK;

	return sched_setattr(pid, &sa, 0);
}

static void set_uclamp_pid(struct uclampset *ctl)
{
	if (ctl->all_tasks) {
		pid_t tid;
		struct proc_tasks *ts = proc_open_tasks(ctl->pid);

		if (!ts)
			err(EXIT_FAILURE, _("cannot obtain the list of tasks"));

		while (!proc_next_tid(ts, &tid))
			if (set_uclamp_one(ctl, tid) == -1)
				err(EXIT_FAILURE, _("failed to set tid %d's uclamp values"), tid);

		proc_close_tasks(ts);

	} else if (set_uclamp_one(ctl, ctl->pid) == -1) {
		err(EXIT_FAILURE, _("failed to set pid %d's uclamp values"), ctl->pid);
	}
}

static void set_uclamp_system(struct uclampset *ctl)
{
	if (!ctl->util_min_set)
		ctl->util_min = read_uclamp_sysfs(_PATH_PROC_UCLAMP_MIN);

	if (!ctl->util_max_set)
		ctl->util_max = read_uclamp_sysfs(_PATH_PROC_UCLAMP_MAX);

	if (ctl->util_min > ctl->util_max) {
		errno = EINVAL;
		err(EXIT_FAILURE, _("util_min must be <= util_max"));
	}

	write_uclamp_sysfs(_PATH_PROC_UCLAMP_MIN, ctl->util_min);
	write_uclamp_sysfs(_PATH_PROC_UCLAMP_MAX, ctl->util_max);
}

static void validate_util(int val)
{
	if (val > 1024 || val < -1) {
		errno = EINVAL;
		err(EXIT_FAILURE, _("%d out of range"), val);
	}
}

int main(int argc, char **argv)
{
	struct uclampset _ctl = {
		.pid = -1,
		.util_min = NOT_SET,
		.util_max = NOT_SET,
		.cmd = NULL
	};
	struct uclampset *ctl = &_ctl;
	int c;

	static const struct option longopts[] = {
		{ "all-tasks",		no_argument, NULL, 'a' },
		{ "pid",		required_argument, NULL, 'p' },
		{ "system",		no_argument, NULL, 's' },
		{ "reset-on-fork",	no_argument, NULL, 'R' },
		{ "help",		no_argument, NULL, 'h' },
		{ "verbose",		no_argument, NULL, 'v' },
		{ "version",		no_argument, NULL, 'V' },
		{ NULL,			no_argument, NULL, 0 }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	close_stdout_atexit();

	while((c = getopt_long(argc, argv, "+asRp:hm:M:vV", longopts, NULL)) != -1)
	{
		switch (c) {
		case 'a':
			ctl->all_tasks = 1;
			break;
		case 'p':
			errno = 0;
			ctl->pid = strtos32_or_err(optarg, _("invalid PID argument"));
			optind++;
			break;
		case 's':
			ctl->system = 1;
			break;
		case 'R':
			ctl->reset_on_fork = 1;
			break;
		case 'v':
			ctl->verbose = 1;
			break;
		case 'm':
			ctl->util_min = strtos32_or_err(optarg, _("invalid util_min argument"));
			ctl->util_min_set = 1;
			validate_util(ctl->util_min);
			optind++;
			break;
		case 'M':
			ctl->util_max = strtos32_or_err(optarg, _("invalid util_max argument"));
			ctl->util_max_set = 1;
			validate_util(ctl->util_max);
			optind++;
			break;
		case 'V':
			print_version(EXIT_SUCCESS);
			/* fallthrough */
		case 'h':
			usage();
		default:
			errtryhelp(EXIT_FAILURE);
		}
	}

	if (argc == 1) {
		usage();
		exit(EXIT_FAILURE);
	}

	/* all_tasks implies --pid */
	if (ctl->all_tasks && ctl->pid == -1) {
		errno = EINVAL;
		err(EXIT_FAILURE, _("missing -p option"));
	}

	if (!ctl->util_min_set && !ctl->util_max_set) {
		/* -p or -s must be passed */
		if (!ctl->system && ctl->pid == -1) {
			usage();
			exit(EXIT_FAILURE);
		}

		show_uclamp_info(ctl);
		return EXIT_SUCCESS;
	}

	/* ensure there's a command to execute if no -s or -p */
	if (!ctl->system && ctl->pid == -1) {
		if (argc <= optind) {
			errno = EINVAL;
			err(EXIT_FAILURE, _("no cmd to execute"));
		}

		argv += optind;
		ctl->cmd = argv[0];
	}

	if (ctl->pid == -1)
		ctl->pid = 0;

	if (ctl->system)
		set_uclamp_system(ctl);
	else
		set_uclamp_pid(ctl);

	if (ctl->verbose)
		show_uclamp_info(ctl);

	if (ctl->cmd) {
		execvp(ctl->cmd, argv);
		errexec(ctl->cmd);
	}

	return EXIT_SUCCESS;
}
